### DISCUSSION TITLE
PHP MySQLI write to database

### DISCUSSION THREAD

**Discussion by ghost** [community/community#136685] [Programming Help]
```php
<?php
class SQLIManager{
    public $connected = false;
    public $connection = null;
    public function open($server, $username, $password){
        if($this->connected === false){
            try{
                $this->connection = mysqli_connect($server, $username, $password);
                if($this->connection->connect_error){
                    throw new Exception("Exception: " .$this->connection->connect_error);
                }else{
                    $this->connected = true;
                }
            }catch(Exception $exception){
                echo $exception;
            }
        }
    }
    public function select($database){
        $db = $this->connection->select_db($database);
        if(!$db){
            die("Error selecting database: " .mysqli_error($this->connection));
        }
    }

    public function write($query, $type, $param){
        $pstm = $this->connection->prepare($query);
        $pstm->bind_param($type, $pstm, $length);
        $pstm->execute();
    }
    public function close(){
        if($this->connected == true){
            $this->connection->close();
            $this->connected = false;
        }
    }
}
```

**Comment by ferid333** [comment]
Great

**Comment by elsycharles** [comment]
There is some issues in your code : 

1. Connection Error Check
    The connection error check is done using $this->connection->connect_error. This is incorrect because mysqli_connect() returns false on failure, not an object. So, accessing connect_error on a failed connection would cause an error.
    here is how you can correct this : 
    `if (!$this->connection) {
    throw new Exception("Exception: " . mysqli_connect_error());
}`

2. Binding Parameters in write() Method
     The method bind_param() is used incorrectly. It requires specific arguments that need to be passed in the correct way. Your current code is passing $pstm and $length which are not defined or correctly used.
     here is how you can correct this : 
     `public function write($query, $type, ...$params){
    $pstm = $this->connection->prepare($query);
    if ($pstm === false) {
        throw new Exception("Exception: " . $this->connection->error);
    }
    $pstm->bind_param($type, ...$params);
    $pstm->execute();
}`

3. Error Handling in select() Method
    The error handling uses mysqli_error($this->connection). If the connection fails, this function might not provide useful information.
    here is how you can correct this : 
    `if (!$this->connection) {
    die("No valid connection to select the database.");
}
if (!$db) {
    die("Error selecting database: " . mysqli_error($this->connection));
}`
    
4. Closing Connection
     The closing of the connection is done correctly, but you could improve readability by using === for comparison.
     here is how you can do it :
     `public function close(){
    if ($this->connected === true) {
        $this->connection->close();
        $this->connected = false;
    }
}`

Hope that can help you.
